vault/resource_pki_secret_backend_cert.go

```diff
@@ -192,16 +205,47 @@ func pkiSecretBackendCertCreate(d *schema.ResourceData, meta interface{}) error
 	d.Set("private_key", resp.Data["private_key"])
 	d.Set("private_key_type", resp.Data["private_key_type"])
 	d.Set("serial_number", resp.Data["serial_number"])
+	d.Set("expiration", resp.Data["expiration"])
 
 	d.SetId(fmt.Sprintf("%s/%s/%s", backend, name, commonName))
 	return pkiSecretBackendCertRead(d, meta)
 }
 
+func pkiSecretBackendCertDiff(d *schema.ResourceDiff, meta interface{}) error {
+	if d.Id() == "" {
+		return nil
+	}
+
+	expiration := d.Get("expiration").(int)
+	expireTime := time.Unix(int64(expiration), 0)
+
+	minSeconds := 0
+	if v, ok := d.GetOk("min_seconds_remaining"); ok {
+		minSeconds = v.(int)
+	}
+
+	renewTime := expireTime.Add(-time.Duration(minSeconds) * time.Second)
+	if time.Now().After(renewTime) {
+		log.Printf("[DEBUG] certificate %q is due for renewal, expires %s, renewal time %s", d.Id(), expireTime, renewTime)
+		d.SetNewComputed("certificate")
+		d.SetNewComputed("private_key")
+		return nil
+	}
+
+	log.Printf("[DEBUG] certificate %q is not due for renewal, expires %s, renewal time %s", d.Id(), expireTime, renewTime)
+	return nil
+}
+
 func pkiSecretBackendCertRead(d *schema.ResourceData, meta interface{}) error {
 	return nil
 }
 
 func pkiSecretBackendCertUpdate(d *schema.ResourceData, m interface{}) error {
+	// If the certificate or private_key have been marked as changed by our custom diff function we need
+	// to create a new certificate and key
+	if d.HasChange("certificate") || d.HasChange("private_key") {
```

Comment:
This code path doesn't appear to have test coverage yet. Would be great to get some here too.